### PR TITLE
Fixed issue with sequences not being poped when Output processor returns Message

### PR DESCRIPTION
**Pull request by sirimamilla**
In AbstractCorrelatingMessageHandler, while processing the outputProcessor result, sequences are not being popped when result is an instance of Message.

#3157

An additional condition is added to handle such sequences.

else if (!(result instanceof Message<?>)) {
					messageBuilder = getMessageBuilderFactory()
							.withPayload(result)
							.copyHeaders(message.getHeaders());
				}
				else if (compareSequences((Message) result, message)) {
					messageBuilder = getMessageBuilderFactory()
							.fromMessage((Message) result);
				}

https://github.com/spring-projects/spring-integration/issues/3157

### CODE REVIEW

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: This is redundant. Since we are going to fail anyway when ` replyChannel.receive(10000)` returns `null` after those 10 seconds.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: Excuse me, and how with this test we can be sure that sequence is popped properly. Doesn't look like other test have some difference with this one, but they don't fails...

**Review by artembilan — CHANGES_REQUESTED** [review]
@sirimamilla ,

thank you for one more contribution!

Here you are some review from me.

Please, add your name to the affected classes.

**Comment by artembilan** [comment]
Also, please, consider to modify a doc on the matter: https://docs.spring.io/spring-integration/docs/5.2.3.RELEASE/reference/html/message-routing.html#aggregatingmessagehandler

> Starting with version 5.1, after processing message group, an AbstractCorrelatingMessageHandler performs a MessageBuilder.popSequenceDetails() message headers modification for the proper splitter-aggregator scenario with several nested levels. It is done only if the message group release result is not a message or collection of messages. 

Also such a breaking change deserves a `What's New` entry. Plus we will add a note in a Migration Guide: https://github.com/spring-projects/spring-integration/wiki/Spring-Integration-5.2-to-5.3-Migration-Guide

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (sirimamilla) [reply spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: HI @artembilan,

When Thread is in hung state due to any issue, the test case is going into hung state. This is safety precaution to ensure test case atleast fails instead of making the whole test suite in hung state forever.

**Comment by sirimamilla** [comment]
> Also, please, consider to modify a doc on the matter: https://docs.spring.io/spring-integration/docs/5.2.3.RELEASE/reference/html/message-routing.html#aggregatingmessagehandler

Where do I modify this doc.. Sorry to bother again. I was not able to rightly identify the Github location to update.

**Comment by artembilan** [comment]
> Where do I modify this doc..

Please, find `aggregator.adoc` file in the project.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (sirimamilla) [reply spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: Added additional condition checks to ensure Sequence is properly popped.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: Something is wrong with your formatting.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: Please, be sure that you don't exceed `120` line length.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: This like is redundant.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: We need a new line in between these methods.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: I would prefer `scatterGatherWireTapChannel`. I mean, please, ensure every word in the name is came-cased. 
Much easier to read.

File: src/reference/asciidoc/aggregator.adoc
Comment (artembilan) [inline src/reference/asciidoc/aggregator.adoc]: Comma before `if`. The `outputMessage` has to be wrapped into a code or words have to be divided.

Also consider to re-phrase the rest of the doc since it confuses now. The text flow has to be smooth somehow and cover together a `Message` and non-`Message` use-cases.

**Review by artembilan — CHANGES_REQUESTED** [review]
Looks like you also have missed to add a `whats-new.adoc` in the `x5.3-general` section.

Thanks

**Comment by sirimamilla** [comment]
Hi @artembilan ,

CI Test cases are all successful in my workstation. but its failing in travis ci. 

> Task :spring-integration-core:testAll
> Task :spring-integration-core:testAll FROM-CACHE
Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.1/userguide/command_line_interface.html#sec:command_line_warnings
BUILD SUCCESSFUL in 29s
19 actionable tasks: 5 executed, 1 from cache, 13 up-to-date

Not sure how to retrigger the CI job.

**Comment by artembilan** [comment]
Yeah... That one fails from time to time.
Not your problem - I've restarted the build on Travis.
Nevertheless it doesn't matter because there is a review from to address or provide your arguments against my points.

Thanks

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (sirimamilla) [reply spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: No changes performed by me here.

Comment (artembilan) [reply spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: Well, GitHub thinks differently and shows me two blank lines.
Only one has to be in between.

Well, don't worry: I clean up the code style on merge: just wanted to let you know that we try to follow some code style. We develop here an Open Source, so our code has to be perfect for community readers.
Therefore, if you are going to follow our rules, we are not going to have discussions like this any more and that will save our time on review and merging process.

Thanks for understanding.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (sirimamilla) [reply spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: I realized that in framework everywhere, wireTap, T is capital. I have fixed this as well.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (sirimamilla) [reply spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: fixed.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (sirimamilla) [reply spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: added additional line break.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (sirimamilla) [reply spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: corrected format

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (artembilan) [reply spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: No reason to comment if you agree with my suggestion for a fix.
I will see your change in the next commit.

I'd like to see comments on review only if you are not agreed with me.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: `sequenceNo` ?
Don't we talk about a `sequenceNumber`?

Consider to use constants from the `IntegrationMessageHeaderAccessor` instead.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (artembilan) [reply spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: This comment has not been addressed.

In some future I'm going to remove this `@Timeout` because I will find the code in the test fully sufficient.
Some comment for this option would save it from removal 😉

File: src/reference/asciidoc/aggregator.adoc
Comment (artembilan) [inline src/reference/asciidoc/aggregator.adoc]: What are these blank lines for?

File: src/reference/asciidoc/aggregator.adoc
Comment (artembilan) [inline src/reference/asciidoc/aggregator.adoc]: Do you see that this is not true any more?
According your logic no one is responsible any more: the framework does that for us already!

Why don't change this text to reflect reality?

File: src/reference/asciidoc/aggregator.adoc
Comment (artembilan) [inline src/reference/asciidoc/aggregator.adoc]: And why `sequenceDetails` is not a code any more?
It can be a plain text, but make it like that then!
I mean like this `sequence details`. Or bring it back to the code wrapping.

File: src/reference/asciidoc/whats-new.adoc
Comment (artembilan) [inline src/reference/asciidoc/whats-new.adoc]: There has to be a blank like after a title.

The title is slightly wrong. Would be better to say: `Aggregator Changes`. Your one doesn't give an clue what I'm to read about.

Also consider to add a like to the chapter. See other sections here for samples.

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java
Comment (sirimamilla) [reply spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: I was under impression grade build throws all issues with code style. But it's not throwing these issues. Next time will ensure manual verification of these

Comment (artembilan) [reply spring-integration-core/src/test/java/org/springframework/integration/dsl/routers/RouterTests.java]: Yeah... The Checkstyle tool is just a tool written by human. Since not all tasks human can perform, it is not a surprise that their tools are not perfect.
More over how can the tool be sure that these two lines is not your intention.
Therefore my question about them: with a good reason we can leave them.
But for great code it would be better to avoid unnecessary changes.

File: src/reference/asciidoc/aggregator.adoc
Comment (sirimamilla) [reply src/reference/asciidoc/aggregator.adoc]: If they are using older version and referring to the latest document,  this will reflect that

File: src/reference/asciidoc/aggregator.adoc
Comment (sirimamilla) [reply src/reference/asciidoc/aggregator.adoc]: I thought it would be better readable with blank lines in between

File: src/reference/asciidoc/aggregator.adoc
Comment (artembilan) [reply src/reference/asciidoc/aggregator.adoc]: Then they have to refer to documentation for that exact version.
The current doc has to explain exactly the current behavior.

Please, don't confuse people! 😄

File: src/reference/asciidoc/aggregator.adoc
Comment (artembilan) [reply src/reference/asciidoc/aggregator.adoc]: Yeah... And why haven't you changed the whole doc for the same style? 😉 
Where is consistency? 

I mean it's OK to have only one line between paragraphs: the doc tool is going to format it for us properly.
More over the final decision about formatting is left on us, project owner. So, you can change something for code style and formatting, but we may reject it.

**Comment by artembilan** [comment]
@sirimamilla ,

any updates here?
Looks like there are some reviews not addressed this or other way. 
We can't merge it without mutual agreement.

We have time though for this until the next week or so.

Thanks

**Comment by sirimamilla** [comment]
@Timeout(11000) is just a safety. If we run the same test case against 5.2.3, Timeout will help to atleast complete the unit test cases with a failure of this test case. if we remove the time out, we may have to abort the complete suit.

**Review by artembilan — APPROVED** [review]
Merging with some clean up...

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/1d9a818efe61a46c6d6a70cc0425e756de765749

@sirimamilla ,

thank you again for contribution!

**Pull request closed without merging**
